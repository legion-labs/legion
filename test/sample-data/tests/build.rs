use std::{
    env,
    path::{Path, PathBuf},
};

use legion_data_compiler::Manifest;

pub fn target_dir() -> PathBuf {
    env::current_exe()
        .ok()
        .map(|mut path| {
            path.pop();
            if path.ends_with("deps") {
                path.pop();
            }
            path
        })
        .unwrap_or_else(|| panic!("cannot find test directory"))
}

pub fn data_build_exe() -> PathBuf {
    target_dir().join(format!("data-build{}", env::consts::EXE_SUFFIX))
}

fn exec_create_build_index(
    path: impl AsRef<Path>,
    project: impl AsRef<Path>,
) -> std::io::Result<std::process::Output> {
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("create");
    command.arg(path.as_ref().to_str().unwrap());
    command.arg(format!("--project={}", project.as_ref().to_str().unwrap()));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

fn exec_data_compile(
    compile_path: &str,
    buildindex_path: &Path,
    destination: &Path,
) -> std::io::Result<std::process::Output> {
    let target = "game";
    let platform = "windows";
    let locale = "en";
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("compile");
    command.arg(compile_path);
    command.arg(format!("--cas={}", destination.to_str().unwrap()));
    command.arg(format!("--target={}", target));
    command.arg(format!("--platform={}", platform));
    command.arg(format!("--locale={}", locale));
    command.arg(format!(
        "--buildindex={}",
        buildindex_path.to_str().unwrap()
    ));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

/// read the build index and remove non-deterministic data.
fn read_build_index(buildindex_path: &Path) -> String {
    let content = std::fs::read_to_string(&buildindex_path).expect("file content");
    let mut content = json::parse(&content).expect("valid json");
    content.remove("project_index");
    content.pretty(2)
}

#[test]
fn incremental_build() {
    let work_dir = tempfile::tempdir().unwrap();
    let sampledata_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let project = sampledata_dir.join("project.index");
    let buildindex = work_dir.as_ref().join("build.index");

    // create build index and do a source pull
    exec_create_build_index(&buildindex, project).expect("new build index");

    insta::assert_snapshot!("initial_index", read_build_index(&buildindex));

    // default root object in sample data
    // /world/sample_1.ent (offline_entity) => runtime_entity
    let root_entity = "d004cd1c00000000fcd3242ec9691beb|019c8223";

    //
    // first data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let manifest: Manifest = serde_json::from_slice(&out.stdout).expect("valid manifest");
    insta::assert_json_snapshot!("first_manifest", manifest);

    let first_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("first_index", first_buildindex);

    //
    // incremental data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let incremental_manifest: Manifest =
        serde_json::from_slice(&out.stdout).expect("valid manifest");

    insta::assert_json_snapshot!("incremental_manifest", incremental_manifest);

    let incremental_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("incremental_index", incremental_buildindex);

    // incremental build with no changes should not affect the build index
    assert_eq!(first_buildindex, incremental_buildindex);
}
