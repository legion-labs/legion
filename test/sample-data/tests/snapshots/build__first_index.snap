---
source: test/sample-data/tests/build.rs
expression: first_buildindex

---
{
  "version": "0.1.0",
  "compiled_resources": [
    {
      "compile_path": "0abe069400000000cea897a15bc29f32|35b5717e",
      "context_hash": "742ad8186628b62d",
      "source_hash": "4f713c07e1efda7f",
      "compiled_path": "0abe069400000000cea897a15bc29f32|35b5717e",
      "compiled_checksum": "eba744d2f9b1fd43f7387be926659fb9",
      "compiled_size": 40
    },
    {
      "compile_path": "0cba171e000000005127bdca42145845|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "75ba9cf2e1383a17",
      "compiled_path": "0cba171e000000005127bdca42145845|30495fed",
      "compiled_checksum": "5c7c0f211147643f6135ba0158d9ddb1",
      "compiled_size": 17
    },
    {
      "compile_path": "0cba171e000000007a80bfadcb6549e7|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e000000007a80bfadcb6549e7|30495fed",
      "compiled_checksum": "da505289290b378da32cd3fb9bf9353a",
      "compiled_size": 17
    },
    {
      "compile_path": "0cba171e000000007df5d0fcdb1eb69b|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e000000007df5d0fcdb1eb69b|30495fed",
      "compiled_checksum": "da505289290b378da32cd3fb9bf9353a",
      "compiled_size": 17
    },
    {
      "compile_path": "0cba171e0000000090b1c112655b00e8|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e0000000090b1c112655b00e8|30495fed",
      "compiled_checksum": "da505289290b378da32cd3fb9bf9353a",
      "compiled_size": 17
    },
    {
      "compile_path": "0cba171e00000000ce1185cf2b953536|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e00000000ce1185cf2b953536|30495fed",
      "compiled_checksum": "da505289290b378da32cd3fb9bf9353a",
      "compiled_size": 17
    },
    {
      "compile_path": "0cba171e00000000e578fadce0648bcc|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e00000000e578fadce0648bcc|30495fed",
      "compiled_checksum": "da505289290b378da32cd3fb9bf9353a",
      "compiled_size": 17
    },
    {
      "compile_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "context_hash": "8854984f90ef7613",
      "source_hash": "7e64701df81d7c9e",
      "compiled_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "compiled_checksum": "f1d27712258c8696e2420a3fb7656288",
      "compiled_size": 64
    },
    {
      "compile_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "1b71a86a85e8968e",
      "compiled_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "compiled_checksum": "22a9232f77613f21445fc27e92e3944e",
      "compiled_size": 200
    },
    {
      "compile_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "83d7786249ffd4e6",
      "compiled_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "compiled_checksum": "bb6a8668f30c94f6a1632390545e7559",
      "compiled_size": 133
    },
    {
      "compile_path": "97b0740f0000000091d021d6f4e05efe|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "6f589089e28fcb6d",
      "compiled_path": "97b0740f0000000091d021d6f4e05efe|aad89045",
      "compiled_checksum": "da8a27471f9b01809ec1f9ba9ebf46f9",
      "compiled_size": 150
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_checksum": "1c8f86eb4e170479f0072a93de78229a",
      "compiled_size": 260
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "compiled_checksum": "3b38a5c712832077652f65948452db76",
      "compiled_size": 2719523
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_Black Background",
      "compiled_checksum": "55e3d72e731d68b86cdbecf6f2d72586",
      "compiled_size": 2500048
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_Ferris",
      "compiled_checksum": "413a76e3a5005efef75be32d7644feb4",
      "compiled_size": 2500634
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo",
      "compiled_checksum": "405ff6a014c8e3008041a0aa1c5719cc",
      "compiled_size": 2500634
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness",
      "compiled_checksum": "405ff6a014c8e3008041a0aa1c5719cc",
      "compiled_size": 2500634
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal",
      "compiled_checksum": "405ff6a014c8e3008041a0aa1c5719cc",
      "compiled_size": 2500634
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee",
      "context_hash": "d821b584c9f0338a",
      "source_hash": "dbd2482a4c49e25b",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness",
      "compiled_checksum": "405ff6a014c8e3008041a0aa1c5719cc",
      "compiled_size": 2500634
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo|c23b950d",
      "context_hash": "db5708d4e77de55c",
      "source_hash": "b4fcee80e405bf11",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo|c23b950d",
      "compiled_checksum": "ac3a5ecc7b49df329a5f8b73ce832a82",
      "compiled_size": 1000000
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness|c23b950d",
      "context_hash": "db5708d4e77de55c",
      "source_hash": "b4fcee80e405bf11",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness|c23b950d",
      "compiled_checksum": "ac3a5ecc7b49df329a5f8b73ce832a82",
      "compiled_size": 1000000
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal|c23b950d",
      "context_hash": "db5708d4e77de55c",
      "source_hash": "b4fcee80e405bf11",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal|c23b950d",
      "compiled_checksum": "ac3a5ecc7b49df329a5f8b73ce832a82",
      "compiled_size": 1000000
    },
    {
      "compile_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness|c23b950d",
      "context_hash": "db5708d4e77de55c",
      "source_hash": "b4fcee80e405bf11",
      "compiled_path": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness|c23b950d",
      "compiled_checksum": "ac3a5ecc7b49df329a5f8b73ce832a82",
      "compiled_size": 1000000
    },
    {
      "compile_path": "e2e9edca000000007714d9ad404557f7|84935d1f",
      "context_hash": "37d0ca216e71d856",
      "source_hash": "4dcf2948669710ae",
      "compiled_path": "e2e9edca000000007714d9ad404557f7|84935d1f",
      "compiled_checksum": "e263e009b1549e1e3bbaf198f2474a64",
      "compiled_size": 897
    }
  ],
  "compiled_resource_references": [
    {
      "compile_path": "0cba171e000000005127bdca42145845|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "75ba9cf2e1383a17",
      "compiled_path": "0cba171e000000005127bdca42145845|30495fed",
      "compiled_reference": "97b0740f0000000062aa07fb54e7a8ee|aad89045"
    },
    {
      "compile_path": "0cba171e000000007a80bfadcb6549e7|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e000000007a80bfadcb6549e7|30495fed",
      "compiled_reference": "97b0740f0000000091d021d6f4e05efe|aad89045"
    },
    {
      "compile_path": "0cba171e000000007df5d0fcdb1eb69b|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e000000007df5d0fcdb1eb69b|30495fed",
      "compiled_reference": "97b0740f0000000091d021d6f4e05efe|aad89045"
    },
    {
      "compile_path": "0cba171e0000000090b1c112655b00e8|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e0000000090b1c112655b00e8|30495fed",
      "compiled_reference": "97b0740f0000000091d021d6f4e05efe|aad89045"
    },
    {
      "compile_path": "0cba171e00000000ce1185cf2b953536|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e00000000ce1185cf2b953536|30495fed",
      "compiled_reference": "97b0740f0000000091d021d6f4e05efe|aad89045"
    },
    {
      "compile_path": "0cba171e00000000e578fadce0648bcc|30495fed",
      "context_hash": "d78e22f18e251521",
      "source_hash": "739114d453b56ff8",
      "compiled_path": "0cba171e00000000e578fadce0648bcc|30495fed",
      "compiled_reference": "97b0740f0000000091d021d6f4e05efe|aad89045"
    },
    {
      "compile_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "context_hash": "8854984f90ef7613",
      "source_hash": "7e64701df81d7c9e",
      "compiled_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "compiled_reference": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo|c23b950d"
    },
    {
      "compile_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "context_hash": "8854984f90ef7613",
      "source_hash": "7e64701df81d7c9e",
      "compiled_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "compiled_reference": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness|c23b950d"
    },
    {
      "compile_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "context_hash": "8854984f90ef7613",
      "source_hash": "7e64701df81d7c9e",
      "compiled_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "compiled_reference": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal|c23b950d"
    },
    {
      "compile_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "context_hash": "8854984f90ef7613",
      "source_hash": "7e64701df81d7c9e",
      "compiled_path": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
      "compiled_reference": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness|c23b950d"
    },
    {
      "compile_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "1b71a86a85e8968e",
      "compiled_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "compiled_reference": "e2e9edca000000007714d9ad404557f7|84935d1f"
    },
    {
      "compile_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "1b71a86a85e8968e",
      "compiled_path": "97b0740f00000000519d4a9f47df0ce5|aad89045",
      "compiled_reference": "e2e9edca000000007714d9ad404557f7|84935d1f"
    },
    {
      "compile_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "83d7786249ffd4e6",
      "compiled_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "compiled_reference": "0cba171e00000000ce1185cf2b953536|30495fed"
    },
    {
      "compile_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "83d7786249ffd4e6",
      "compiled_path": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
      "compiled_reference": "0cba171e00000000e578fadce0648bcc|30495fed"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "0abe069400000000cea897a15bc29f32|35b5717e"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "0cba171e000000005127bdca42145845|30495fed"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "0cba171e000000007a80bfadcb6549e7|30495fed"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "0cba171e000000007df5d0fcdb1eb69b|30495fed"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "0cba171e0000000090b1c112655b00e8|30495fed"
    },
    {
      "compile_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "context_hash": "8360eb38f405f86f",
      "source_hash": "37f12d7a953e6498",
      "compiled_path": "97b0740f00000000fcd3242ec9691beb|aad89045",
      "compiled_reference": "97b0740f00000000519d4a9f47df0ce5|aad89045"
    },
    {
      "compile_path": "e2e9edca000000007714d9ad404557f7|84935d1f",
      "context_hash": "37d0ca216e71d856",
      "source_hash": "4dcf2948669710ae",
      "compiled_path": "e2e9edca000000007714d9ad404557f7|84935d1f",
      "compiled_reference": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a"
    }
  ],
  "pathid_mapping": {
    "0abe069400000000cea897a15bc29f32": "0abe069400000000cea897a15bc29f32",
    "0cba171e000000005127bdca42145845": "0cba171e000000005127bdca42145845",
    "0cba171e000000007a80bfadcb6549e7": "0cba171e000000007a80bfadcb6549e7",
    "0cba171e000000007df5d0fcdb1eb69b": "0cba171e000000007df5d0fcdb1eb69b",
    "0cba171e0000000090b1c112655b00e8": "0cba171e0000000090b1c112655b00e8",
    "0cba171e00000000ce1185cf2b953536": "0cba171e00000000ce1185cf2b953536",
    "0cba171e00000000e578fadce0648bcc": "0cba171e00000000e578fadce0648bcc",
    "0f6f8b3a000000007b21d6d843b4b7a2": "1793ed6400000000f8b1bd1dceaf6f33|0f6f8b3a",
    "1793ed6400000000f8b1bd1dceaf6f33": "1793ed6400000000f8b1bd1dceaf6f33",
    "30495fed000000000f002d85caa5ad62": "0cba171e00000000e578fadce0648bcc|30495fed",
    "30495fed000000001ca4261976e4b3c5": "0cba171e000000007df5d0fcdb1eb69b|30495fed",
    "30495fed000000008e21cbbe5c7a6256": "0cba171e000000007a80bfadcb6549e7|30495fed",
    "30495fed00000000cd581e368043ab83": "0cba171e00000000ce1185cf2b953536|30495fed",
    "30495fed00000000f8e5bce43227fd8d": "0cba171e0000000090b1c112655b00e8|30495fed",
    "30495fed00000000f9dbba18688917d1": "0cba171e000000005127bdca42145845|30495fed",
    "332b6ce800000000d4b05b1a98368098": "332b6ce800000000d4b05b1a98368098",
    "35b5717e0000000084a0d1047b638b96": "0abe069400000000cea897a15bc29f32|35b5717e",
    "84935d1f000000007f3751314a9b2cb0": "e2e9edca000000007714d9ad404557f7|84935d1f",
    "97b0740f00000000519d4a9f47df0ce5": "97b0740f00000000519d4a9f47df0ce5",
    "97b0740f0000000062aa07fb54e7a8ee": "97b0740f0000000062aa07fb54e7a8ee",
    "97b0740f0000000091d021d6f4e05efe": "97b0740f0000000091d021d6f4e05efe",
    "97b0740f00000000fcd3242ec9691beb": "97b0740f00000000fcd3242ec9691beb",
    "aad890450000000007e9c88e6e870796": "97b0740f0000000062aa07fb54e7a8ee|aad89045",
    "aad89045000000004b9a8871d48cd4a1": "97b0740f00000000519d4a9f47df0ce5|aad89045",
    "aad8904500000000ab5fe63eda1e1c4f": "97b0740f00000000fcd3242ec9691beb|aad89045",
    "aad8904500000000e07de9e7a949885e": "97b0740f0000000091d021d6f4e05efe|aad89045",
    "c23b950d000000000c3544b040d370c4": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo|c23b950d",
    "c23b950d0000000018f62f13833ca491": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness|c23b950d",
    "c23b950d000000006bbebf2a63fe287c": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal|c23b950d",
    "c23b950d00000000c507b5e792c2c22d": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness|c23b950d",
    "cf00bfee0000000016aa3dee7b5442b6": "d8a1e0c20000000061f6de1be4764697|cf00bfee_normal",
    "cf00bfee0000000024c5940da4439d36": "d8a1e0c20000000061f6de1be4764697|cf00bfee_albedo",
    "cf00bfee000000002efb3622525057e3": "d8a1e0c20000000061f6de1be4764697|cf00bfee_roughness",
    "cf00bfee00000000a911bfb22afdcd7d": "d8a1e0c20000000061f6de1be4764697|cf00bfee_metalness",
    "d8a1e0c20000000061f6de1be4764697": "d8a1e0c20000000061f6de1be4764697",
    "e2e9edca000000007714d9ad404557f7": "e2e9edca000000007714d9ad404557f7"
  }
}
